runtime/caml/camlatomic.h

```diff
@@ -56,6 +57,31 @@ typedef _Atomic intnat atomic_intnat;
 #define atomic_store_relaxed(p, v)                      \
   atomic_store_explicit((p), (v), memory_order_relaxed)
 
+Caml_inline void caml_atomic_counter_init(atomic_uintnat* counter, uintnat n)
+{
+  atomic_store_release(counter, n);
+}
+
+/* Decrement an atomic uintnat counter. Assertion check for
+ * underflow. Returns the new value. */
+
+Caml_inline uintnat caml_atomic_counter_decr(atomic_uintnat* counter)
+{
+  uintnat old = atomic_fetch_sub(counter, 1);
+  CAMLassert ((intnat)old > 0);
```

Comment:
Noob question, sorry: why do we cast from `uintnat` to `intnat` before performing this comparison? (And also `old+1` below.)

Reply:
Well we cannot test if an unsigned number is negative… but we could test for `old != CAML_UINTNAT_MAX`, which means the counter would have looped (counter was 0).

Reply:
My understanding is that we are testing that the pre-decrement value was *strictly positive* (not zero), which also makes sense on unsigned integers.

Reply:
Good catch; I originally wrote this test differently, as a test on the new value (which, as @MisterDA notes, can never be negative) and the cast survived a tweak. I'm removing both casts.